src/Riok.Mapperly/Descriptors/MappingBuilders/MemoryMappingBuilder.cs

```diff
@@ -0,0 +1,236 @@
+
+            // if source is ReadOnlyMemory and target is Memory
+            // either use ToArray or MapToSpan(memory.Span)
+            (CollectionType.ReadOnlyMemory, CollectionType.Memory) => BuildMemoryToArrayMapping(ctx, elementMapping),
+
+            // if target is Memory
+            // either use Cast or MapToArray(memory.Span)
+            (_, CollectionType.ReadOnlyMemory or CollectionType.Memory) => BuildMemoryToMemoryMapping(ctx, elementMapping),
+
+            // if target is Array
+            // either use ToArray or MapToArray(memory.Span)
+            (_, CollectionType.Array) => BuildMemoryToArrayMapping(ctx, elementMapping),
+
+            // if target is Enumerable
+            // use MapToEnumerable(memory.Span)
+            (_, CollectionType.Enumerable) => BuildMemoryToEnumerableMapping(ctx, elementMapping),
+
+            _ => throw new Exception("Unreachable exception.")
+        };
+    }
+
+    public static IExistingTargetMapping? TryBuildExistingTargetMapping(MappingBuilderContext ctx)
+    {
+        if (!ctx.IsConversionEnabled(MappingConversionType.Memory))
+            return null;
+
+        if (GetCollectionInfo(ctx.Source, ctx) is not { } sourceInfo)
+            return null;
+
+        if (GetCollectionInfo(ctx.Target, ctx) is not { } targetInfo)
+            return null;
+
+        if (!IsMemory(sourceInfo.TypeOfCollection) && !IsMemory(targetInfo.TypeOfCollection))
+            return null;
+
+        // can only map to Enumerable. Existing types Span,Memory and array are all immutable
+        if (targetInfo.TypeOfCollection is not CollectionType.Enumerable)
+        {
+            ctx.ReportDiagnostic(DiagnosticDescriptors.CannotMapToReadOnlyMember, ctx.Target);
+            return null;
+        }
+
+        if (ctx.FindOrBuildMapping(sourceInfo.EnumeratedType, targetInfo.EnumeratedType) is not { } elementMapping)
+            return null;
+
+        var sourceSpan = ctx.Types.Get(typeof(ReadOnlySpan<>)).Construct(elementMapping.SourceType);
+        if (ctx.FindOrBuildExistingTargetMapping(sourceSpan, ctx.Target) is not ExistingTargetMapping enumerableMapping)
+            return null;
+
+        return new SourceObjectMemberDelegateExistingTargetMapping(ctx.Source, ctx.Target, SpanMemberName, enumerableMapping);
+    }
+
+    private static TypeMapping? BuildSpanToMemoryMapping(MappingBuilderContext ctx, ITypeMapping elementMapping)
+    {
+        if (!elementMapping.IsSynthetic || ctx.MapperConfiguration.UseDeepCloning)
+        {
+            var targetArray = ctx.Types.GetArrayType(elementMapping.TargetType);
+            if (ctx.FindOrBuildMapping(ctx.Source, targetArray.NonNullable()) is not { } arrayMapping)
+                return null;
+            return new CastMapping(ctx.Source, ctx.Target, arrayMapping);
+        }
+
+        return new SourceObjectMethodDelegateMapping(ctx.Source, ctx.Target, ToArrayMethodName);
```

Comment:
IMO the readability of this method could be improved if the first if is inverted and returned early and the second one is changed into `?:` syntax.